### ISSUE TITLE
replaceReducer eliminates enhancer

### ISSUE THREAD

**Issue by antontsvil** [reduxjs/redux#4127]
### What is the current behavior?
If you use an enhancer when creating your Redux store, the first subsequent call to `replaceReducer` on the store seems to remove the enhancer.

### Steps to Reproduce
[Run this JSFiddle and observe the console](https://jsfiddle.net/antontsvil/tcn9rm7z/3/)
The enhancer will stop logging actions following the call to `replaceReducer`

### What is the expected behavior?
Not sure if what I'm reporting is not a bug and I'm misunderstanding the usage of enhancers; But seems to me enhancers should persist for the lifetime of the store, regardless of which reducers are being used.

**Comment by timdorr** [comment]
That isn't an enhancer. It's just a wrapper function on `createStore`. To create an enhancer, you would need to change something about the store itself, such as replacing the `replaceReducer` function on the resulting store.

**Comment by markerikson** [comment]
Mmm... it looks like a valid enhancer in terms of _syntax_, at least. But yeah, the real point of an enhancer is to override one of the actual store methods. All this is doing is wrapping the original reducer function.  So yes, if you replace that reducer, you've... replaced the reducer :)

**Comment by Methuselah96** [comment]
This would actually be a bug in the enhancer. If you're enhancer messes with the reducer, it needs to re-implement the `replaceReducer` function. See https://github.com/reduxjs/redux/pull/3776/files/b6483652e8d68663bf355061bc2b34a63c09e978#diff-326d6e3ed223aeeb0f5f2099b79d5910860494677628d63c13b2ff3bb223ae3eL14 for an example.

**Comment by Methuselah96** [comment]
P.S. @timdorr @markerikson It would be great to get https://github.com/reduxjs/redux/pull/3776 merged.

**Comment by Methuselah96** [comment]
@antontsvil Here's a fiddle showing how it should be implemented: https://jsfiddle.net/m3hcs84r/.

**Comment by markerikson** [comment]
@Methuselah96 the issue with those PRs is that they're against `master`, which is the still unreleased "convert all of Redux to TS" codebase, which we also have no timeline to release atm.

In order to actually be releasable any time soon, we'd need updates made to the TS typedefs in the `4.x` branch.

**Comment by Methuselah96** [comment]
@markerikson I understand, I can re-implement https://github.com/reduxjs/redux/pull/3776 for `4.x` (https://github.com/reduxjs/redux/pull/3772 reverts a change that's only in 5.x). What are the breaking changes in 5.x? Are the types different enough that it needs a new major?

**Comment by markerikson** [comment]
That's sorta the issue - we're _not_ sure what the breaking changes are, it's likely the types would benefit from _more_ changes, and there's not enough reason to push it live and cause churn in the ecosystem atm :)

The conversion work was actually done in... late 2019, I think?  And it's been sitting there in `master` ever since.

Basically, the community contributed the work after we asked for it, and then the whole effort just stalled.

There's also some tie-in to `redux-thunk`'s types being stuck in a holding pattern as well.

Basically, what's live in 4.x right now works well enough, and especially with RTK now being the default, that there just isn't enough pressing reason to try to push through a 5.0 release any time soon.

**Comment by Methuselah96** [comment]
Yeah, makes sense. Sorry to derail this issue. FWIW I have annoying casts in my code that could be removed if https://github.com/reduxjs/redux-thunk/pull/255 were released. It's frustrating seeing completed work from 2 years ago that fixes issues that I'm dealing with that has no timeline for a release. (I could potentially augment the interface, but I was kinda hoping it would just get released at some point.) These situations make DT seem a lot more appealing so the types lifecycle don't have to be as tightly coupled to the JS releases, but I understand that decision has a lot of pros/cons and has been hashed and rehashed many times and frankly I'm not sure which I prefer.

But I also understand that keeping churn down is valuable, so thanks for the hard work in that regard.

**Comment by Methuselah96** [comment]
(But I'm also not on RTK yet, as noted in https://github.com/reduxjs/redux-thunk/issues/299.)

**Comment by markerikson** [comment]
Yeah, I'd like to see 5.0 released too, but it needs some meaningful attention to figure out what other types changes might be necessary, and given everything else going on it's just at the bottom of the priority list for us right now.

If someone else with serious TS knowledge wants to help us out and try to push the current TS port forward by evaluating it and seeing where things stand and how it can be improved, I'd love it.

**Comment by markerikson** [comment]
Opened up #4129 to cover that topic.

**Comment by antontsvil** [comment]
Thanks for pointing out what I figured was the initial issue, was that I wasn't understanding enhancers 100%. And thank you for your helpful solution @Methuselah96. I do have to note that I had to modify it because your solution never actually uses `nextReducer`. Which means that the enhancer will in fact continue to run, but operate with the old reducer. [What I've done to resolve this](https://replit.com/@AntonTsvilikhov/Redux-Enhancer-Blaster#index.js) is pass the new reducer to the `enhancedReducer` method and use the initial reducer as a fallback.

This seems to fix my issue (enhancer which supports replaceReducer AND operates using the new reducer) but I'm not sure if it's an antipattern, looks a bit weird tbh. Just leaving the solution here for anyone else encountering the same problem.
